sources/metricdog/src/systemd.rs

```diff
@@ -0,0 +1,270 @@
+use crate::service_check::{ServiceCheck, ServiceHealth};
+
+use crate::error::{self, Result};
+use crate::host_check::HostCheck;
+use log::trace;
+use snafu::{OptionExt, ResultExt};
+use std::process::Command;
+
+const EXIT_STATUS_PROPERTY: &str = "ExecMainStatus";
+/// This systemd unit property records the time between boot and when the unit enters the 'active'
+/// state in the order of microseconds
+const ACTIVE_TIMESTAMP_PROPERTY: &str = "ActiveEnterTimestampMonotonic";
+
+const SYSTEMCTL: &str = "/usr/bin/systemctl";
+const JOURNALCTL: &str = "/usr/bin/journalctl";
+
+#[derive(Clone, Copy)]
+pub(crate) struct SystemdCheck {}
+
+impl ServiceCheck for SystemdCheck {
+    fn check(&self, service_name: &str) -> Result<ServiceHealth> {
+        if is_ok(service_name)? {
+            return Ok(ServiceHealth {
+                is_healthy: true,
+                exit_code: None,
+            });
+        }
+        Ok(ServiceHealth {
+            is_healthy: false,
+            exit_code: parse_service_exit_code(service_name)?,
+        })
+    }
+}
+
+struct Outcome {
+    exit: i32,
+    stdout: String,
+}
+
+impl Outcome {
+    fn is_exit_true(&self) -> bool {
+        self.exit == 0
+    }
+}
+
+fn command(cmd: &str, args: &[&str]) -> Result<Outcome> {
+    trace!("calling '{}' with '{:?}'", cmd, args);
+    let output = Command::new(cmd)
+        .args(args)
+        .output()
+        .with_context(|_| error::CommandSnafu {
+            command: cmd,
+            args: args.iter().map(|&s| s.to_owned()).collect::<Vec<String>>(),
+        })?;
+    Ok(Outcome {
+        exit: output.status.code().unwrap_or(-1),
+        stdout: String::from_utf8_lossy(output.stdout.as_slice()).into(),
+    })
+}
+
+fn is_active(service: &str) -> Result<bool> {
+    let outcome = command(SYSTEMCTL, &["is-active", service])?;
+    Ok(outcome.is_exit_true())
+}
+
+fn is_failed(service: &str) -> Result<bool> {
+    let outcome = command(SYSTEMCTL, &["is-failed", service])?;
+    Ok(outcome.is_exit_true())
+}
+
+fn is_ok(service: &str) -> Result<bool> {
+    Ok(!is_failed(service)? && is_active(service)?)
+}
+
+/// Parses out the systemd unit's exit status from `systemctl show` output
+fn parse_service_exit_code(service: &str) -> Result<Option<i32>> {
+    // we don't check the command's exit code because systemctl returns non-zero codes for various
+    // non-exceptional execution outcomes.
+    let outcome = command(
+        SYSTEMCTL,
+        &["show", "--property", EXIT_STATUS_PROPERTY, service],
+    )?;
+    Ok(parse_property(&outcome.stdout, EXIT_STATUS_PROPERTY)
+        .and_then(|exit_code| exit_code.parse::<i32>().ok()))
+}
+
+impl HostCheck for SystemdCheck {
+    fn is_first_boot(&self) -> Result<bool> {
+        // Upon first boot, we expect this output to have a single line entry leading with '0 <hash>'
+        let outcome = command(JOURNALCTL, &["--list-boots", "-q"])?;
```

Comment:
nit: I prefer not to mix short and long options. The long ones tend to be more self-documenting in that I don't have to go to a manual or run the command to confirm that it does what I expect.

```suggestion
        let outcome = command(JOURNALCTL, &["--list-boots", "--quiet"])?;
```